lang/rust/patches/01-add_mips64_musl_support.patch

```diff
@@ -0,0 +1,30 @@
+--- /dev/null
++++ b/compiler/rustc_target/src/spec/mips64_openwrt_linux_musl.rs
+@@ -0,0 +1,16 @@
++use crate::spec::{Target, TargetOptions};
++
++pub fn target() -> Target {
++    let mut base = super::linux_musl_base::opts();
++    base.cpu = "mips64r2".to_string();
++    base.features = "+mips64r2,+soft-float".to_string();
++    base.max_atomic_width = Some(64);
++    Target {
++        // LLVM doesn't recognize "muslabi64" yet.
++        llvm_target: "mips64-unknown-linux-musl".to_string(),
++        pointer_width: 64,
++        data_layout: "E-m:e-i8:8:32-i16:16:32-i64:64-n32:64-S128".to_string(),
++        arch: "mips64".to_string(),
++        options: TargetOptions { endian: "big".to_string(), mcount: "_mcount".to_string(), ..base },
++    }
```

Comment:
This looks useful to other distros too :) have you considered upstreaming it to rust-lang/rust?

Reply:
I don't see why not, although it's just the built-in triple renamed and defined with a soft-float since the dynlinking is handles in the MUSL define.

Reply:
Is there any progress with upstreaming patches which you have for mips64, aarch64, etc?

Reply:
`mips64-openwrt-linux-musl` was upstream'ed and merged into `rust-lang` master back in February, 2022.  It's the only actual hardware I have to test on, architect-wise.

As I continue with testing, or getting confirmation that the various tuples don't cause issues (`armv5` still doesn't work, for example) and figuring out whether errors I see are because of OpenWrt build system, the tuple, the rust install, or the package itself (or some other interaction and/or combined situations).  For example, I've had a request for the `vaultwarden` package.  I was seeing errors, I reported errors, and in the end, it was because the `ring` crate (which is a common crypto crate from what I can tell) uses `BoringSSL`, and `BoringSSL` doesn't work for BigEndian archs (like `mips64`) *sigh*.  So chasing errors to ensure stability is a long, drawn out process.

And, even when everything seems to *work*, the end product may or may not work as intended.  I had issues with `suricata6` in which it would build with `soft-float` flagged, enabled in the `mips64-openwrt-linux-musl` tuple, but was SIGSEV erroring when I attempted to run it.  Turns out, OpenWrt for `mips64` didn't enable MIPS `soft-float` emulation in the kernel.

So, if anyone wants to help me test an arch, verification can happen and I can upstream them.  They are responsive enough.